### PR TITLE
Checkpoint refactors

### PR DESCRIPTION
**Pull request by czue**
can be read commit-by-commit. codependent with https://github.com/dimagi/commcare-hq/pull/8892

I think this is pretty safe, though open to push-back that I should put on staging for a bit. Have tested the different pillow-types locally

### CODE REVIEW

**Comment by czue** [comment]
wait for build on https://github.com/dimagi/commcare-hq/pull/8892 but I think it should be fixed now

**Comment by czue** [comment]
(@snopoke @proteusvacuum)

**Comment by czue** [comment]
:white_check_mark:

File: pillowtop/listener.py [hunk 1]
```diff
@@ -320,15 +321,14 @@ def wait_expired(self):
...
             self.change_queue.append(change)
             if self.queue_full or self.wait_expired:
                 self.process_chunk()
         elif self.python_filter(change['doc']) or (change.get('deleted', None) and self.process_deletions):
             self.process_change(change)
-        if self.changes_seen % self.checkpoint_frequency == 0 and do_set_checkpoint:
+        if context.changes_seen % self.checkpoint_frequency == 0 and context.do_set_checkpoint:
             # if using chunking make sure we never allow the checkpoint to get in
             # front of the chunks
             if self.use_chunking:
```
Comment (snopoke) [inline pillowtop/listener.py]: Should `self.set_checkpoint(change)` still be called here (2 lines below)

Comment (snopoke) [inline pillowtop/listener.py]: `self.process_chunk()` could also be called from `fire_change_processed_event`

File: pillowtop/pillow/interface.py [hunk 2]
```diff
@@ -65,24 +70,42 @@ def process_changes(self, since, forever):
...
                         notify_exception(None, u'processor error in pillow {} {}'.format(
                             self.get_name(), e,
                         ))
                         raise
+                    else:
+                        self.fire_change_processed_event(change, context)
                 else:
                     self.checkpoint.touch(min_interval=CHECKPOINT_MIN_WAIT)
         except PillowtopCheckpointReset:
-            self.changes_seen = 0
```
Comment (snopoke) [inline pillowtop/pillow/interface.py]: Should this not reset the context?

File: pillowtop/listener.py (hunk 1 above)
Comment (czue) [inline pillowtop/listener.py]: yeah i didn't bother changing `PythonPillow` at all yet but am planning to soon

File: pillowtop/pillow/interface.py (hunk 2 above)
Comment (czue) [inline pillowtop/pillow/interface.py]: the recursive call should reset it at the beginning of the function

**Pull request merged**